python-frontend/src/main/java/org/sonar/python/caching/Caching.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.caching;
+
+import com.google.protobuf.InvalidProtocolBufferException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.sonar.plugins.python.api.caching.CacheContext;
+import org.sonar.python.index.AmbiguousDescriptor;
+import org.sonar.python.index.ClassDescriptor;
+import org.sonar.python.index.Descriptor;
+import org.sonar.python.index.DescriptorUtils;
+import org.sonar.python.index.FunctionDescriptor;
+import org.sonar.python.index.VariableDescriptor;
+import org.sonar.python.types.protobuf.DescriptorsProtos;
+
+public class Caching {
+
+  private final CacheContext cacheContext;
+
+  public static final String INPORT_MAP_CACHE_KEY_PREFIX = "python_import_map:";
+  public static final String PROJECT_SYMBOL_TABLE_CACHE_KEY_PREFIX = "python_project_symbol_table:";
+
+  public Caching(CacheContext cacheContext) {
+    this.cacheContext = cacheContext;
+  }
+
+  public void writeImportMapEntry(String moduleFqn, Set<String> imports) {
+    byte[] importData = String.join(";", imports).getBytes(StandardCharsets.UTF_8);
+    String cacheKey = INPORT_MAP_CACHE_KEY_PREFIX + moduleFqn;
+    cacheContext.getWriteCache().write(cacheKey, importData);
+  }
+
+  public void writeProjectLevelSymbolTableEntry(String moduleFqn, Set<Descriptor> descriptors) {
+    String cacheKey = PROJECT_SYMBOL_TABLE_CACHE_KEY_PREFIX + moduleFqn;
+    cacheContext.getWriteCache().write(cacheKey, moduleDescriptor(descriptors).toByteArray());
+   }
+
+  public Optional<Set<Descriptor>> readProjectLevelSymbolTableEntry(String moduleFqn) {
+    cacheContext.getReadCache().contains(moduleFqn);
+    byte[] bytes = cacheContext.getReadCache().readBytes(moduleFqn);
+    if (bytes == null) {
+      return Optional.empty();
+    }
+    try {
+      return Optional.of(DescriptorUtils.deserializeProtobufDescriptors(bytes));
+    } catch (InvalidProtocolBufferException e) {
+      e.printStackTrace();
```

Comment:
I think we may want to re-throw an exception here.

Reply:
Actually I think we can log something here and return null. The caller will know the cache entry is missing/corrupted through the null return.